test/ha/ha.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ha
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/url"
+	"strings"
+	"testing"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	restclient "k8s.io/client-go/rest"
+	pkgTest "knative.dev/pkg/test"
+	rtesting "knative.dev/serving/pkg/testing/v1"
+	"knative.dev/serving/test"
+	"knative.dev/serving/test/e2e"
+	v1test "knative.dev/serving/test/v1"
+)
+
+const (
+	servingNamespace = "knative-serving"
+	haReplicas       = 2
+)
+
+func getLeader(t *testing.T, clients *test.Clients, labelSelector string) (string, error) {
+	var leader string
+	if err := wait.PollImmediate(test.PollInterval, time.Minute, func() (bool, error) {
+		controllerPodList, err := clients.KubeClient.Kube.CoreV1().Pods(servingNamespace).List(metav1.ListOptions{
+			LabelSelector: labelSelector,
+		})
+		if err != nil {
+			return false, fmt.Errorf("error retrieving pods with label %s: %w", labelSelector, err)
+		}
+		for _, pod := range controllerPodList.Items {
+			podLogs, err := logsAsString(clients.KubeClient.Kube.CoreV1().Pods(servingNamespace).GetLogs(pod.Name, &corev1.PodLogOptions{}))
+			if err != nil {
+				if strings.Contains(err.Error(), "ContainerCreating") {
+					// Need to wait a little more for the containers to be started
+					return false, nil
```

Comment:
should it rather be continue here to watch the second pod logs?

Reply:
That's a good point.